chain/src/consume_unverified.rs

```diff
@@ -1,405 +1,245 @@
-    }
-
-    fn check_if_hardfork_during_blocks(
-        &self,
-        hardfork: &HardForks,
-        blocks: &VecDeque<BlockView>,
-    ) -> bool {
-        if blocks.is_empty() {
-            false
-        } else {
-            // This method assumes that the input blocks are sorted and unique.
-            let rfc_0049 = hardfork.ckb2023.rfc_0049();
-            let epoch_first = blocks.front().unwrap().epoch().number();
-            let epoch_next = blocks
-                .back()
-                .unwrap()
-                .epoch()
-                .minimum_epoch_number_after_n_blocks(1);
-            epoch_first < rfc_0049 && rfc_0049 <= epoch_next
-        }
-    }
-}
+    unverified_block_rx: Receiver<UnverifiedBlock>,
+    truncate_block_rx: Receiver<TruncateRequest>,
 
-pub(crate) struct GlobalIndex {
-    pub(crate) number: BlockNumber,
-    pub(crate) hash: Byte32,
-    pub(crate) unseen: bool,
+    stop_rx: Receiver<()>,
+    processor: ConsumeUnverifiedBlockProcessor,
 }
 
-impl GlobalIndex {
-    pub(crate) fn new(number: BlockNumber, hash: Byte32, unseen: bool) -> GlobalIndex {
-        GlobalIndex {
-            number,
-            hash,
-            unseen,
+impl ConsumeUnverifiedBlocks {
+    pub(crate) fn new(
+        shared: Shared,
+        unverified_blocks_rx: Receiver<UnverifiedBlock>,
+        truncate_block_rx: Receiver<TruncateRequest>,
+        proposal_table: ProposalTable,
+        verify_failed_blocks_tx: tokio::sync::mpsc::UnboundedSender<VerifyFailedBlockInfo>,
+        stop_rx: Receiver<()>,
+    ) -> Self {
+        ConsumeUnverifiedBlocks {
+            tx_pool_controller: shared.tx_pool_controller().to_owned(),
+            unverified_block_rx: unverified_blocks_rx,
+            truncate_block_rx,
+            stop_rx,
+            processor: ConsumeUnverifiedBlockProcessor {
+                shared,
+                proposal_table,
+                verify_failed_blocks_tx,
+            },
         }
     }
 
-    pub(crate) fn forward(&mut self, hash: Byte32) {
-        self.number -= 1;
-        self.hash = hash;
-    }
-}
-
-/// Chain background service
-///
-/// The ChainService provides a single-threaded background executor.
-pub struct ChainService {
-    shared: Shared,
-    proposal_table: ProposalTable,
-}
-
-impl ChainService {
-    /// Create a new ChainService instance with shared and initial proposal_table.
-    pub fn new(shared: Shared, proposal_table: ProposalTable) -> ChainService {
-        ChainService {
-            shared,
-            proposal_table,
-        }
-    }
+    pub(crate) fn start(mut self) {
+        loop {
+            let _trace_begin_loop = minstant::Instant::now();
+            select! {
+                recv(self.unverified_block_rx) -> msg => match msg {
+                    Ok(unverified_task) => {
+                        // process this unverified block
+                        trace!("got an unverified block, wait cost: {:?}", begin_loop.elapsed());
```

Comment:
`begin_loop` seems to be undefined. Is this a typo (should have been `_trace_begin_loop`)?